docs/adr/0003-consensus-runtime-token-transfer.md

```diff
@@ -0,0 +1,460 @@
+
+### State
+
+This proposal introduces/updates the following consensus state in the staking
+module:
+
+#### General Accounts
+
+The general account data structure is modified to include an additional field
+storing the allowances as follows:
+
+```golang
+type GeneralAccount struct {
+    // ... existing fields omitted ...
+
+    Approvals map[Address]*quantity.Quantity `json:"approvals,omitempty"`
+}
+```
+
+### Transaction Methods
+
+This proposal adds the following new transaction methods in the staking module:
+
+#### Approve
+
+Approve enables an account holder to set an allowance for a beneficiary.
+
+**Method name:**
+
+```
+staking.Approve
+```
+
+**Body:**
+
+```golang
+type Approve struct {
+    Beneficiary  Address           `json:"beneficiary"`
+    Negative     bool              `json:"negative,omitempty"`
+    AmountChange quantity.Quantity `json:"amount_change"`
+}
+```
+
+**Fields:**
+
+- `beneficiary` specifies the beneficiary account address.
+- `amount_change` specifies the absolute value of the amount of base units to
+  change the allowance for.
+- `negative` specifies whether the `amount_change` should be subtracted instead
+  of added.
+
+The transaction signer implicitly specifies the general account. Upon executing
+the approval the following actions are performed:
+
+- If the `disable_allowances` staking consensus parameters is set to `true`,
+  the method fails with `ErrForbidden`.
+
+- The account indicated by the signer is loaded.
+
+- If the approval would create a new allowance and the maximum number of
+  allowances for an account has been reached, the method fails with
+  `ErrInvalidArgument`.
+
+- The set of approvals is updated so that the allowance is updated as specified
+  by `amount_change`/`negative`. In case the change would cause the allowance to
+  be equal to zero or negative, the allowance is removed.
+
+- The account is saved.
+
+- A corresponding `ApprovalEvent` is emitted with the following structure:
+
+  ```golang
+  type ApprovalEvent struct {
+      Owner       Address           `json:"owner"`
+      Beneficiary Address           `json:"beneficiary"`
+      Allowance   quantity.Quantity `json:"allowance"`
+  }
+  ```
+
+  Where `allowance` contains the new total allowance. The event is emitted even
+  if the new allowance is zero.
+
+#### Withdraw
+
+Withdraw enables a beneficiary to withdraw from the given account.
+
+**Method name:**
+
+```
+staking.Withdraw
+```
+
+**Body:**
+
+```golang
+type Withdraw struct {
+    From   Address           `json:"from"`
+    Amount quantity.Quantity `json:"amount"`
+}
+```
+
+**Fields:**
+
+- `from` specifies the account address to withdraw from.
+- `amount` specifies the amount of base units to withdraw.
+
+The transaction signer implicitly specifies the destination general account.
+Upon executing the withdrawal the following actions are performed:
+
+- If either the `disable_transfers` or `disable_allowances` staking consensus
```

Comment:
this, whereas Approve doesn't check `disable_transfers`. but what would be the purpose of having approves without withdraws? maybe we should use the same criteria for both.

Reply:
Agreed.

Reply:
Updated.